subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -187,11 +196,24 @@ private static void walkNestedChildIfPresentOrElse(Supplier<Object> unpacker, St
             }
             if (value != null) {
                 handler.accept(value);
+            } else if (isElementOfCollection) {
+                throw new IllegalStateException(getNullNestedCollectionValueExceptionMessage(qualifiedName));
             } else {
-                missingHandler.run();
+                TypeToken<?> getterType = TypeToken.of(propertyMetadata.getGetterMethod().getGenericReturnType());
+                TypeMetadata typeMetadata = extractNestedTypeFromGetter(getterType);
+                visitor.visitNested(typeMetadata, qualifiedName, propertyMetadata, null);
             }
         }
 
+        @SuppressWarnings("unchecked")
+        private TypeMetadata extractNestedTypeFromGetter(TypeToken<?> typeToken) {
+            if (!Provider.class.isAssignableFrom(typeToken.getRawType())) {
+                return getTypeMetadata(typeToken.getRawType());
+            }
+            TypeToken<?> nestedToken = extractNestedType((TypeToken<Provider<?>>) typeToken, Provider.class, 0);
+            return extractNestedTypeFromGetter(nestedToken);
+        }
```

Comment:
:x: This method could be improved in a few ways. Its name suggests that it's a friend of  `extractNestedType()`, but it actually isn't.

The important bit here is to _unpack_ the type. I'd do that like this:

```java
        private static Class<?> unpackType(Class<?> type) {
            while (Provider.class.isAssignableFrom(type)) {
                type = extractNestedType(TypeToken.of((Class<? extends Provider<?>>) type), Provider.class, 0).getRawType();
            }
            return type;
        }
```

...and call `getTypeMetadata()` at the call-site instead.

Reply:
Changed. But TypeToken has to be used, otherwise we don't have parameter information.

Reply:
Ah, yes.